File: soothfast-bind/tests/csharp_smoke/Smoke.cs
using Acme.Core;

var counter = new Counter(10);
Console.WriteLine($"value={counter.Value}");
Console.WriteLine($"bump={counter.Bump(5)}");
Console.WriteLine($"at(Low)={counter.At(Level.Low)}");
Console.WriteLine($"at(High)={counter.At(Level.High)}");
Console.WriteLine($"bumpAll={counter.BumpAll(new long[] { 1, 2, 3 })}");
try
{
    counter.Bump(long.MaxValue);
    throw new Exception("expected CoreException");
}
catch (CoreException e)
{
    Console.WriteLine($"caught: {e.Message}");
}
counter.Dispose();

Console.WriteLine($"digest=[{string.Join(", ", Core.Digest(new byte[] { 1, 2, 3 }))}]");
Console.WriteLine($"normalize=[{string.Join(", ", Core.Normalize(new double[] { 1.0, 2.0, 3.0 }, 2.0))}]");
Console.WriteLine($"stamp={Core.Stamp(7, 3.0, new byte[] { 9, 9 })}");

var scaled = new double[] { 0, 0, 0 };
Core.ScaleInto(new double[] { 1.0, 2.0, 3.0 }, 2.0, scaled);
Console.WriteLine($"scaleInto=[{string.Join(", ", scaled)}]");

Console.WriteLine($"peakLevel={Core.PeakLevel(new double[] { 0.1, 0.9, 0.3 })}");
var found = Core.FindCounter(5);
Console.WriteLine($"findCounter={found.Value}");
Console.WriteLine($"findCounter(missing)={Core.FindCounter(-1)?.Value.ToString() ?? "null"}");

Console.WriteLine($"describe={Core.Describe("world")}");
Console.WriteLine($"describe(none)={Core.Describe(null) ?? "null"}");

Console.WriteLine($"describeOwned={Core.DescribeOwned("world")}");
Console.WriteLine($"describeOwned(none)={Core.DescribeOwned(null) ?? "null"}");

try
{
    Core.Fail("bad byte");
    throw new Exception("expected CoreException");
}
catch (CoreException e)
{
    Console.WriteLine($"fail caught: {e.Message}");
}

var lo = new double[3];
var hi = new double[3];
Core.Split(new double[] { 1.0, 2.0, 3.0 }, lo, hi);
Console.WriteLine($"split.lo=[{string.Join(", ", lo)}]");
Console.WriteLine($"split.hi=[{string.Join(", ", hi)}]");

var closedCounter = new Counter(1);
closedCounter.Dispose();
try
{
    _ = closedCounter.Value;
    throw new Exception("expected ObjectDisposedException");
}
catch (ObjectDisposedException e)
{
    Console.WriteLine($"use after close: {e.ObjectName}");
}

File: soothfast-bind/tests/cpp/Smoke.cpp
#include "core.hpp"

#include <cstdint>
#include <cstdio>
#include <limits>
#include <vector>

using namespace acme::core;

int main() {
    Counter c(10);
    std::printf("value=%lld\n", static_cast<long long>(c.value()));
    std::printf("at(Low)=%lld\n", static_cast<long long>(c.at(Level::Low)));
    std::printf("at(High)=%lld\n", static_cast<long long>(c.at(Level::High)));
    std::printf("bump=%lld\n", static_cast<long long>(c.bump(5)));

    std::vector<int64_t> by{1, 2, 3};
    std::printf("bumpAll=%lld\n", static_cast<long long>(c.bump_all(by)));

    Counter overflowing(std::numeric_limits<int64_t>::max());
    try {
        overflowing.bump(1);
        std::printf("no exception thrown\n");
    } catch (const Error &e) {
        std::printf("caught: %s\n", e.what());
    }

    std::vector<uint8_t> data{1, 2, 3};
    auto digested = digest(data);
    std::printf("digest=[%d, %d, %d]\n", digested[0], digested[1], digested[2]);

    std::vector<double> input{1.0, 2.0, 3.0};
    auto normalized = normalize(input, 2.0);
    std::printf("normalize=[%g, %g, %g]\n", normalized[0], normalized[1], normalized[2]);

    std::printf("greet=%s\n", greet("C++").c_str());

    std::vector<double> out(3);
    scale_into(input, 2.0, out);
    std::printf("scale_into=[%g, %g, %g]\n", out[0], out[1], out[2]);

    std::printf("stamp=%llu\n", static_cast<unsigned long long>(stamp(1, 2.0, data)));

    std::vector<double> peaked{0.1, 0.9, 0.3};
    std::printf("peak_level=%s\n", peak_level(peaked) == Level::High ? "High" : "Low");

    auto found = find_counter(5);
    std::printf("find_counter=%lld\n", static_cast<long long>(found->value()));
    std::printf("find_counter(missing)=%s\n", find_counter(-1).has_value() ? "present" : "none");

    std::printf("describe=%s\n", describe("world").value().c_str());
    std::printf("describe(none)=%s\n", describe(std::nullopt).has_value() ? "present" : "none");

    std::printf("describe_owned=%s\n", describe_owned("world").value().c_str());
    std::printf("describe_owned(none)=%s\n", describe_owned(std::nullopt).has_value() ? "present" : "none");

    try {
        fail("bad byte");
        std::printf("fail: no exception thrown\n");
    } catch (const Error &e) {
        std::printf("fail caught: %s\n", e.what());
    }

    std::vector<double> src{1.0, 2.0, 3.0};
    std::vector<double> lo(3);
    std::vector<double> hi(3);
    split(src, lo, hi);
    std::printf("split.lo=[%g, %g, %g]\n", lo[0], lo[1], lo[2]);
    std::printf("split.hi=[%g, %g, %g]\n", hi[0], hi[1], hi[2]);

    return 0;
}
